src/copy.ml

```diff
@@ -1037,28 +1031,18 @@ let transferFileUsingExternalCopyprog
     else
       Prefs.read copyprog
   in
-  let extraquotes = Prefs.read copyquoterem = `True
-                 || (  Prefs.read copyquoterem = `Default
-                    && Util.findsubstring "rsync" prog <> None) in
-  let addquotes root s =
-    match root with
-    | Common.Local, _ -> s
-    | Common.Remote _, _ -> if extraquotes then Uutil.quotes s else s in
+  let progTokens = Str.split (Str.regexp "[ \t]+") prog in
   let fromSpec =
       (formatConnectionInfo rootFrom)
-    ^ (addquotes rootFrom
-         (Fspath.toString (Fspath.concat (snd rootFrom) pathFrom))) in
+    ^ (Fspath.toString (Fspath.concat (snd rootFrom) pathFrom)) in
   let toSpec =
       (formatConnectionInfo rootTo)
-    ^ (addquotes rootTo
-         (Fspath.toString (Fspath.concat fspathTo pathTo))) in
-  let cmd = prog ^ " "
-             ^ (Uutil.quotes fromSpec) ^ " "
-             ^ (Uutil.quotes toSpec) in
-  Trace.log (Printf.sprintf "%s\n" cmd);
+    ^ (Fspath.toString (Fspath.concat fspathTo pathTo)) in
+  Trace.log (prog ^ " " ^ fromSpec ^ " " ^ toSpec ^ "\n");
   Lwt_util.resize_region !copyprogReg (Prefs.read copymax);
   Lwt_util.run_in_region !copyprogReg 1
-    (fun () -> External.runExternalProgram cmd) >>= fun (_, log) ->
+    (fun () -> External.runExternalProgramArgs (List.hd progTokens)
+                 (Array.of_list (List.append (List.tl progTokens) [fromSpec; toSpec]))) >>= fun (_, log) ->
```

Comment:
Here I would suggest a style change.
Instead of using `List.hd`, `List.tl` and `List.append`, you could write
```ocaml
  let prog :: args = Str.split (Str.regexp "[ \t]+") prog in (* or whatever names you choose instead of prog and args *)
  Lwt_util.run_in_region !copyprogReg 1
    (fun () -> External.runExternalProgramArgs prog
                 (Array.of_list (args @ [fromSpec; toSpec]))) >>= fun (_, log) ->

Reply:
Thanks! `@` was something I was unable to find with quite a lot of searching.

With this change, I now get a warning of a non-exhaustive match, as the list could be empty. I looked into this possibility before, and it seems that the list can't in fact be empty, because if `copyprog` is set to an empty value, then the default is used. I assume there is some idiom for this case?

Reply:
> With this change, I now get a warning of a non-exhaustive match, as the list could be empty.

Right, of course. And this is why the code should be written in this style in the first place; to get the compiler to point out the mistakes.
Using, `List.hd` and `List.tl` is worse because both of these can raise runtime exceptions.
The better way would then be something like:
```ocaml
  let prog, args (* or whatever names you choose instead of prog and args *) = match Str.split (Str.regexp "[ \t]+") prog with [] -> assert false (* still fails at runtime *) | h :: t -> h, t in
  Lwt_util.run_in_region !copyprogReg 1
    (fun () -> External.runExternalProgramArgs prog
                 (Array.of_list (args @ [fromSpec; toSpec]))) >>= fun (_, log) ->
```

As you can see, this can still fail at runtime. If this is unacceptable then validation must be performed to make sure that the original value of `prog` is never empty. Maybe this validation already exists, I haven't checked.

Reply:
What is the correct way of specifying args? Right now you exclude the executable itself from the list of args, but isn't that supposed to be the first argument? (that is, instead of extracting the tail of list, just use the entire list)

Reply:
> As you can see, this can still fail at runtime. If this is unacceptable then validation must be performed to make sure that the original value of `prog` is never empty. Maybe this validation already exists, I haven't checked.

As I said above, `prog` can't be empty, because if it is, the default value is used. So I think it would be sufficient to add a comment to this effect.

Reply:
> What is the correct way of specifying args? Right now you exclude the executable itself from the list of args, but isn't that supposed to be the first argument? (that is, instead of extracting the tail of list, just use the entire list)

That didn't occur to me; in any case, it doesn't appear to be the case, because the code works as-is.

The documentation is ambiguous; `open_process_args` refers the reader to `open_process_args_out`, which refers the reader to `open_process_args_in` (a rather eccentric ordering IMO: why isn't `open_process_args` first in the documentation, with the other two functions referring directly to it? and there are other problems, as `open_process_args_out` is not the same as `open_process_args_in` except for redirecting stdin to a pipe, as it doesn't redirect stdout to a pipe…). Anyway (I shall file an upstream bug report!). `open_process_args_in` says it "runs the program `prog` with arguments `args`. This is the bit that's ambiguous. As I say, from the fact that the code I wrote works, it clearly doesn't allow `argv[0]` to be set differently from `prog`, but this is not clear.

Reply:
> > As you can see, this can still fail at runtime. If this is unacceptable then validation must be performed to make sure that the original value of `prog` is never empty. Maybe this validation already exists, I haven't checked.
> 
> As I said above, `prog` can't be empty, because if it is, the default value is used. So I think it would be sufficient to add a comment to this effect.

It still can be empty. All it takes is for someone to change the default to empty (unlikely, but still) or just passing an whitespace-only value from command line. It's a minor issue and the added validation would still have to terminate the program which will also happen with `assert false`, except in the latter case you can't specify the error message. This is not a problem for me.

> > What is the correct way of specifying args? Right now you exclude the executable itself from the list of args, but isn't that supposed to be the first argument? (that is, instead of extracting the tail of list, just use the entire list)
> 
> That didn't occur to me; in any case, it doesn't appear to be the case, because the code works as-is.

All these functions work by calling `posix_spawn` or `execve`. The man pages have this to say:
for execve(2):
```
By convention, the first of these strings (i.e., argv[0]) should contain the filename associated with the file being executed.
```
for posix_spawn(3):
```
The arguments to the created process come from the argv argument, which is processed as for execve(2).
```

I don't know what is correct...

Reply:
Sorry, I crossed paths with you above and came to the same conclusion; I will now revisit my PR.

Reply:
Oh, of course my code seemed to work because the default value for `copyprog` contains multiple options, so the first of these became `argv[0]`.

Reply:
I've updated the PR to check the list is non-empty (with an `assert false`), and to use all the `args` as arguments.

Reply:
I have filed https://github.com/ocaml/ocaml/pull/12105 to try to improve the documentation of `open_process_args*`.